compiler/rustc_infer/src/infer/error_reporting/suggest.rs

```diff
@@ -351,6 +351,59 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
         }
     }
 
+    pub(super) fn suggest_function_pointers(
+        &self,
+        cause: &ObligationCause<'tcx>,
+        span: Span,
+        exp_found: &ty::error::ExpectedFound<Ty<'tcx>>,
+        diag: &mut Diagnostic,
+    ) {
+        debug!("suggest_function_pointers(cause={:?}, exp_found={:?})", cause, exp_found);
+        let ty::error::ExpectedFound { expected, found } = exp_found;
+        let expected_inner = expected.peel_refs();
+        let found_inner = found.peel_refs();
+        if !expected_inner.is_fn() || !found_inner.is_fn() {
+            return;
+        }
+        diag.note("different `fn` items have unique types, even if their signatures are the same");
```

Comment:
We shouldn't emit this note if there are two fn pointers.

Reply:
Also, ideally, we'd special case the note for fn ptr vs item, since there's only one fn item in that context. Maybe mention that fn items are distinct from fn pointers..

Reply:
for fn pointers, should we emit a note that fn pointers are distinct, or just return?